### PR TITLE
2429/single clamav

### LINKED ISSUE #2429: Singular ClamAV Scanner
**Description:**
To reduce consumed RAM, we want to collapse the 3 clamav instances into one. For ATO reasons, the prod backend must use an instance in the prod space. 

**Acceptance Criteria:** 
- [ ] Point all our backend apps instance to a tanf-prod clamav scanner
- [ ] Ensure all pipeline/deployment automation reinforces this new structure
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
- [ ] Update backend code specific to clamav location
- [ ] Update deployment scripts which link/unlink (both deploy-backend and circleci)
- [ ] Ensure space-level networking is opened such that all staging/dev apps can speak with prod scanner
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
- This should free up approximately 4GB of RAM usage across the org and might prevent future deployment failures due to memory quota ceiling.

**Supporting Documentation:**

**Open Questions:**

### PR DESCRIPTION
**Pull request by George-Hudson**
## Summary of Changes
In order to preserve cf space in dev and staging, we are reducing the number of ClamAV instances we are using. Only the ClamAV instance in the prod space will be used. We also created NGINX servers to act as gatekeepers for each space and to forward requests to the production ClamAV server.

Pull request closes #2429

## How to Test
- the ClamAV scan needs function as intended when running through the gatekeeper.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] [**ClamAV scans still work appropriately**]
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

**Review by andrew-jameson — APPROVED** [review]
deploy-backend.sh:L43, we can delete AV_SCAN_URL now, please do so, otherwise LGTM!

File: scripts/deploy-backend.sh [hunk 1]
```diff
@@ -107,6 +106,17 @@ update_backend()
...
 
+    if ["$CF_SPACE" = "tanf-prod" ]; then
+      # Add network policy to allow backend to access tanf-prod services
+      cf add-network-policy "$CGAPPNAME_BACKEND" clamav-rest --protocol tcp --port 9000
+      cf set-env "$CGAPPNAME_BACKEND" AV_SCAN_URL "http://tanf-prod-clamav-rest.apps.internal:9000/scan"
+    else
+      cf add-network-policy "$CGAPPNAME_BACKEND" tdp-clamav-nginx-$env --protocol tcp --port 9000
+
+      # Add environment varilables for clamav
+      cf set-env "$CGAPPNAME_BACKEND" AV_SCAN_URL "https://tdp-clamav-nginx-$env.apps.internal:9000/scan"
```
Comment (jtimpe) [inline scripts/deploy-backend.sh]: does this require a different nginx proxy in each environment?

**Review by jtimpe — APPROVED** [review]
lgtm! just one question for clarification

File: scripts/deploy-backend.sh (hunk 1 above)
Comment (raftmsohani) [reply scripts/deploy-backend.sh]: Yes, we basically have two nginx proxy, one for dev and one for staging

Comment (jtimpe) [reply scripts/deploy-backend.sh]: are those deployed in ci or manually via the instructions included in the readme?

**Review by raftmsohani — APPROVED** [review]
LGTM!

**Review by elipe17 — APPROVED** [review]

**Comment by ADPennington** [comment]
@George-Hudson @raftmsohani getting the following result in logs when submitting a file:
clamav_qasp

```
17:01:20.841: [APP/PROC/WEB.0] 2023-10-05 21:01:20,840 ERROR clients.py::scan_file:L74 :  ClamAV connection failure: HTTPSConnectionPool(host='tdp-clamav-nginx-dev.apps.internal', port=9000): Max retries exceeded with url: /scan (Caused by SSLError(SSLError(1, '[SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:997)')))
17:01:20.841: [APP/PROC/WEB.0] ClamAV connection failure: HTTPSConnectionPool(host='tdp-clamav-nginx-dev.apps.internal', port=9000): Max retries exceeded with url: /scan (Caused by SSLError(SSLError(1, '[SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:997)')))
```

**Review by ADPennington — APPROVED** [review]
looks great @George-Hudson @raftmsohani 🚀 

_file scan via nginx router in tdp-backend-qasp logs_
clam2

_file being scanned in prod clamav instance_
clam1

_scan results saved in DAC in qasp env only (also verified that prod apps unaffected by scans in dev env_
clam3

**Pull request merged**